### PR TITLE
Adding supply credentials for a third-party registry when using Terragrunt Provider Cache

### LINKED ISSUE #3298: Add a way to supply credentials for a third-party registry when using Terragrunt Provider Cache
## Describe the bug
Would like to use Terragrunt provider cache feature with `run-all`. We execute Terragrunt workloads exclusively in Github Actions CI/CD pipelines. Since it is non-interactive we cannot use commands like `terraform login` to auth with third-party provider registries.

We have the following `.terraformrc` configuration file:
```hcl
provider_installation {
    direct {
        exclude = ["registry.terraform.io/*/*"]
    }
    filesystem_mirror {
        path = "hcl-file-working-directory-on-the-ci-cd-runner/terraform.d/plugins/"
        include = ["registry.terraform.io/manually-downloaded/provider"] 
    }
    network_mirror {
        url = "https://myorg.jfrog.io/artifactory/api/terraform/terraform-registry-prod/providers/" #<-- REQUIRES CREDENTIALS
        exclude = ["registry.terraform.io/manually-downloaded/provider"]
    }
}
```
The way we supply credential for `myorg.jfrog.io` is by setting env var: `TF_TOKEN_myorg_jfrog_io: ${{secrets.JFROG_API_KEY}}` which is automatically recognized by Terraform. However, the problem arises when attempting to enable Terragrunt Provider Cache, by setting following vars:
`TERRAGRUNT_PROVIDER_CACHE=1`
`TERRAGRUNT_PROVIDER_CACHE_REGISTRY_NAMES=myorg.jfrog.io`
We get the following error:
image

Terragrunt is unable to pass through credentials specified in `TF_TOKEN_myorg_jfrog_io`

## Expected behavior
Terragrunt is able to read credentials from `TF_TOKEN_` variables.
OR
Introduce new env var and cli flag to supply credentials when using Terragrunt Provider Cache

## Additional Context
We run around 15 Terragrunt workloads in the pipeline at the same time. Each of those calls around 20 different terraform modules using `run-all` feature of Terragrunt. This results in having to download mostly same providers 300 times which is extremely inefficient as well as overwhelming for provider registry server resulting in provider querying errors. It would be absolutely amazing to have Terragrunt Provider Cache working 😊
## Versions
- Terragrunt version: 0.63.6
- Terraform version: 1.5.7
- Environment details: alpine linux container in Github Actions runner.

## Related Issues
https://github.com/gruntwork-io/terragrunt/issues/3287
https://github.com/gruntwork-io/terragrunt/issues/3278

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

Fixes #3298.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

File: terraform/cache/services/provider_cache.go [hunk 1]
```diff
@@ -226,6 +243,26 @@ func (cache *ProviderCache) warmUp(ctx context.Context) error {
...
+	}
+
+	if cache.credsSource == nil {
+		return req, nil
+	}
+
+	hostname := svchost.Hostname(req.URL.Hostname())
+	if creds, err := cache.credsSource.ForHost(hostname); err != nil {
+		return nil, err
+	} else if creds != nil {
```
Comment (denis256) [inline terraform/cache/services/provider_cache.go:259]: I think `else` can be omitted here, since if condition have return

Comment (levkohimins) [reply terraform/cache/services/provider_cache.go]: In fact, the error returned is always `nil`, so I get rid of the error altogether (improved coed).

**Review by denis256 — APPROVED** [review]

**Comment by levkohimins** [comment]
Resolved in [v0.64.2](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.64.2) release.

**Comment by levkohimins** [comment]
@denis256, thanks for the review!

**Pull request merged**